File: src/Archive/CompressionService.cs
namespace Pyxelze;

internal static class CompressionService
{
    public static bool CompressDirectory(string dirPath, string? outputFile = null)
    {
        outputFile ??= Path.Combine(Path.GetDirectoryName(dirPath) ?? "", Path.GetFileName(dirPath) + ".png");
        Logger.Log($"CompressDirectory: {dirPath} -> {outputFile}");

        var pass = PassphrasePrompt.Prompt(
            L.Get("compression.passphraseTitle"),
            L.Get("compression.passphrasePrompt"));
        if (pass == null) return false;

        var passArg = string.IsNullOrEmpty(pass) ? "" : $" {PassphraseManager.BuildPassphraseArg(pass)}";
        var psi = RoxRunner.CreateRoxProcess($"encode \"{dirPath}\" \"{outputFile}\"{passArg}");

        int exit = ProcessHelper.RunWithProgress(
            L.Get("compression.encoding"),
            L.Get("compression.encodingFile", Path.GetFileName(dirPath)),
            psi, out var stdout, out var stderr);

        if (exit == 0 && File.Exists(outputFile))
            return true;

        if (AntivirusHelper.IsAccessDenied(stderr))
            return CompressViaTempFile(dirPath, outputFile, passArg);

        return false;
    }

    private static bool CompressViaTempFile(string dirPath, string outputFile, string passArg)
    {
        var tempFile = Path.Combine(Path.GetTempPath(), $"pyxelze-encode-{Guid.NewGuid():N}.png");
        Logger.Log($"CompressViaTempFile: {dirPath} -> {tempFile}");

        var psi = RoxRunner.CreateRoxProcess($"encode \"{dirPath}\" \"{tempFile}\"{passArg}");

        int exit = ProcessHelper.RunWithProgress(
            L.Get("compression.fallback"),
            L.Get("compression.fallbackProgress"),
            psi, out _, out _);

        if (exit == 0 && File.Exists(tempFile))
        {
            try
            {
                if (File.Exists(outputFile)) File.Delete(outputFile);
                File.Move(tempFile, outputFile);
                return true;
            }
            catch (Exception ex)
            {
                Logger.Log($"CompressViaTempFile move failed: {ex}");
            }
        }

        TempHelper.SafeDeleteFile(tempFile);
        return false;
    }
}

File: src/archive/CompressionService.cpp
#include "CompressionService.h"
#include "core/RoxRunner.h"
#include "core/ProcessHelper.h"
#include "core/Logger.h"
#include "security/PassphraseManager.h"
#include "helpers/TempHelper.h"
#include "ui/PassphraseDialog.h"
#include "ui/ProgressDialog.h"
#include "localization/Localization.h"
#include <QFileInfo>
#include <QDir>
#include <QFile>
#include <QUuid>

static bool isAccessDenied(const QString& err) {
    return err.contains(QStringLiteral("access denied"), Qt::CaseInsensitive)
        || err.contains(QStringLiteral("os error 5"), Qt::CaseInsensitive);
}

static bool compressViaTempFile(QWidget* parent, const QString& dirPath, const QString& outputFile, const QString& passArg) {
    auto tempFile = QDir::tempPath() + QStringLiteral("/pyxelze-encode-%1.png").arg(QUuid::createUuid().toString(QUuid::Id128));
    Logger::log(QStringLiteral("CompressViaTempFile: %1 -> %2").arg(dirPath, tempFile));

    auto args = QStringLiteral("encode \"%1\" \"%2\"%3").arg(dirPath, tempFile, passArg);
    auto r = ProgressDialog::runRoxWithProgress(parent, L::get("compression.fallback"), L::get("compression.fallbackProgress"), args);

    if (r.exitCode == 0 && QFile::exists(tempFile)) {
        QFile::remove(outputFile);
        if (QFile::rename(tempFile, outputFile))
            return true;
    }
    TempHelper::safeDeleteFile(tempFile);
    return false;
}

bool CompressionService::compressDirectory(QWidget* parent, const QString& dirPath, const QString& outputFile) {
    auto out = outputFile;
    if (out.isEmpty()) {
        QFileInfo fi(dirPath);
        out = fi.absolutePath() + QStringLiteral("/") + fi.fileName() + QStringLiteral(".png");
    }
    Logger::log(QStringLiteral("CompressDirectory: %1 -> %2").arg(dirPath, out));

    auto pass = PassphraseDialog::prompt(parent, L::get("compression.passphraseTitle"), L::get("compression.passphrasePrompt"));
    if (!pass.has_value()) return false;

    QString passArg;
    if (!pass->isEmpty())
        passArg = QStringLiteral(" ") + PassphraseManager::buildPassphraseArg(*pass);

    auto args = QStringLiteral("encode \"%1\" \"%2\"%3").arg(dirPath, out, passArg);
    auto r = ProgressDialog::runRoxWithProgress(parent, L::get("compression.encoding"), L::get("compression.encodingFile").replace(QStringLiteral("{0}"), QFileInfo(dirPath).fileName()), args);

    if (r.exitCode == 0 && QFile::exists(out))
        return true;

    if (isAccessDenied(r.stdErr))
        return compressViaTempFile(parent, dirPath, out, passArg);

    return false;
}
